gotham/src/router/tree/node.rs

```diff
@@ -129,24 +130,47 @@ impl Node {
     ///
     /// In the situation where all these avenues are exhausted an InternalServerError will be
     /// provided.
-    pub fn select_route(&self, state: &State) -> Result<&Box<Route + Send + Sync>, StatusCode> {
-        match self.routes.iter().find(|r| r.is_match(state).is_ok()) {
-            Some(route) => {
+    pub fn select_route<'a>(
+        &'a self,
+        state: &State,
+    ) -> Result<&'a Box<Route + Send + Sync>, RouteNonMatch> {
+        enum Selection<'b> {
+            Init,
+            Match(&'b Box<Route + Send + Sync>),
+            NonMatch(RouteNonMatch),
+        }
+
+        let iter = self.routes.iter();
+        let selection = iter.fold(Selection::Init, |s, r| match s {
```

Comment:
Thoughts on making this a `for` loop (with mutable accumulator for `Selection::NonMatch`) so we can short circuit on a `Selection::Match` result?

Reply:
Alternatively it seems like IterTools has a `fold_while` function, which is basically the same thing, https://docs.rs/itertools/*/itertools/trait.Itertools.html#method.fold_while but I'm not sure on adding another dependency just for that.

Reply:
@smangelsdorf did you have any thoughts here? Figured this comment slipped through the cracks.

Reply:
Sorted. It had slipped under a pile of other work, though I was always going to come back to it (eventually 😛).